metricbeat/module/mongodb/dbstats/dbstats.go

```diff
@@ -0,0 +1,90 @@
+package dbstats
+
+import (
+	"errors"
+
+	"github.com/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/metricbeat/mb"
+	"gopkg.in/mgo.v2"
+)
+
+// init registers the MetricSet with the central registry.
+// The New method will be called after the setup of the module and before starting to fetch data
+func init() {
+	if err := mb.Registry.AddMetricSet("mongodb", "dbstats", New); err != nil {
+		panic(err)
+	}
+}
+
+// MetricSet type defines all fields of the MetricSet
+// As a minimum it must inherit the mb.BaseMetricSet fields, but can be extended with
+// additional entries. These variables can be used to persist data or configuration between
+// multiple fetch calls.
+type MetricSet struct {
+	mb.BaseMetricSet
+	dialInfo *mgo.DialInfo
+}
+
+// New create a new instance of the MetricSet
+// Part of new is also setting up the configuration by processing additional
+// configuration entries if needed.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	dialInfo, err := mgo.ParseURL(base.HostData().URI)
+	if err != nil {
+		return nil, err
+	}
+	dialInfo.Timeout = base.Module().Config().Timeout
+
+	return &MetricSet{
+		BaseMetricSet: base,
+		dialInfo:      dialInfo,
+	}, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right format
+// It returns the event which is then forward to the output. In case of an error, a
+// descriptive error must be returned.
+func (m *MetricSet) Fetch() ([]common.MapStr, error) {
+
+	// establish connection to mongo
+	session, err := mgo.DialWithInfo(m.dialInfo)
```

Comment:
@ruflin 

This follows the pattern used by the `status` metricset [here](https://github.com/elastic/beats/blob/master/metricbeat/module/mongodb/status/status.go#L47), but I'm not sure it makes sense to open and close the mongo connection each time fetch() is called.

According to the [docs](https://www.elastic.co/guide/en/beats/metricbeat/current/metricset-details.html#_timeout_connections_to_services) connections should be established in the `New()` method and persist between fetches. 

I have two proposals:

1. Each metricset establishes its own set of connections for fetching data from mongo, and is implemented in New(). The advantage is that each metricset has its own set of connections which can be tailored for collecting the specific metric. The disadvantages are duplicative code and increased resource utilization (although minimal). 

2. In the mongodb package, implement a NewConnection() singleton that is available to both metricsets. The advantage here is reducing the number of connections and resource utilization. The disadvantage is that both metricsets will share the same connection parameters. Therefore, connection parameters cannot be tailored for each metricset individually. Nonetheless, I'm in favor of this method. 

Lastly, based on this implementation I'm more confident that the mongodb module is not connecting directly to the individual instances, but would instead connect to the cluster. This means that any cluster member may answer the query. This issue is a bit more involved, and I don't know enough about the platform to be sure, so I may need to jump on a Zoom call or something to talk further.


P.S. I'm going through such a proposal process as I'm just getting started in Beats. I'm interested in your thoughts and recommendations so I can better understand the philosophy and guiding principals of the platform. If you're preference is for me to dive right into implementation, I'm happy to do so!